platform/lib/routers/pages.js

```diff
@@ -17,173 +17,169 @@
+ * @return {URL}    The eventually rewritten URL
  */
-function getFilteredFormat(request) {
-  const QUERY_PARAMETER_NAME = 'format';
-  const ALLOWED_FORMATS = ['websites', 'stories', 'ads', 'email'];
-
-  const activeFormat = request.query[QUERY_PARAMETER_NAME] || 'websites';
-  if (ALLOWED_FORMATS.indexOf(activeFormat.toLowerCase()) == -1) {
-    // If the format to filter by is invalid or none use websites
-    return 'websites';
+function ensureUrlScheme(originalUrl) {
+  const url = new URL(originalUrl, config.hosts.platform.base);
+
+  // Get rid of former .amp.html file extension for legacy support
+  if (url.pathname.endsWith('.amp.html')) {
+    url.pathname = url.pathname.slice(0, -9);
   }
 
-  return activeFormat;
-}
+  // Get rid of .html file extension
+  if (url.pathname.endsWith('.html')) {
+    url.pathname = url.pathname.slice(0, -5);
+  }
 
-/**
- * Checks if a path ends on a directory and appends index.html if that's
- * the case, otherwise appends .html extension
- * @param  {String} filePath
- * @return {String}
- */
-function ensureFileExtension(filePath) {
-  if (filePath.endsWith('/')) {
-    return filePath += 'index.html';
+  // Get rid of index in the URL
+  if (url.pathname.endsWith('index')) {
+    url.pathname = url.pathname.slice(0, -5);
   }
 
-  const extension = path.extname(filePath);
-  if (!extension) {
-    return filePath += '.html';
+  // Ensure there is a trailing slash
+  if (!url.pathname.endsWith('/')) {
+    url.pathname = `${url.pathname}/`;
   }
 
-  return filePath;
+  return url;
 }
 
+// Used to speed up resolving of path stubs to valid paths
+const pathCache = new LRU({
+  max: 500,
+});
 
-// Setup a proxy over to Grow during development
-if (config.isDevMode()) {
-  // Only import the stuff needed for proxying during development
-  const modifyResponse = require('http-proxy-response-rewrite');
-  const got = require('got');
-  const {pageTransformer} = require('@lib/build/pageTransformer');
+/**
+ * Fetches a template matching the requested path
+ * @param  {String}             templatePath The path where the template can be found
+ * @return {nunjucks.Template|null}
+ */
+async function loadTemplate(templatePath) {
+  let template = null;
+
+  // The path has been ensured to always have a trailing slash which isn't
+  // needed to find a matching page file
+  templatePath = templatePath.slice(0, -1);
+
+  const resolvedPath = pathCache.get(templatePath);
+  // If the path has already been tried to resolve but never found
+  // do not try to resolve it again
+  if (resolvedPath === false) {
+    return template;
+  }
 
-  // Also create a logger during development since you want to know
-  // what's going on
-  const log = new Signale({
-    'interactive': false,
-    'scope': 'Grow (Proxy)',
-  });
+  // As the request path is not the actual path to the template it is somehow
+  // guessed by testing all of AVAILABLE_STUBS ...
+  for (const stub of AVAILABLE_STUBS) {
```

Comment:
Moving this part into a method makes it easy to implement this without `continue` and `break` making it a lot easier to follow the control flow.